runtime/runtime/src/verifier.rs

```diff
@@ -131,6 +135,40 @@ pub fn validate_transaction(
         .map_err(|_| InvalidTxError::CostOverflow.into())
 }
 
+/// Processes a group of transactions for the same (account, public_key) in ascending nonce order,
+/// using a frozen snapshot for read‑only access.
+/// Returns a Vec of (SignedTransaction, VerificationResult).
+pub fn verify_and_charge_transactions(
+    config: &RuntimeConfig,
+    state_update: &mut TrieUpdate,
+    group: &[(SignedTransaction, TransactionCost)],
+    block_height: Option<BlockHeight>,
+    current_protocol_version: ProtocolVersion,
+    frozen_snapshot: &FrozenMemTries,
+) -> Result<Vec<(SignedTransaction, VerificationResult)>, InvalidTxError> {
+    if group.is_empty() {
+        return Ok(vec![]);
+    }
+    let first_tx = &group[0].0;
+    let signer_id = first_tx.transaction.signer_id();
+    let public_key = first_tx.transaction.public_key();
+
+    let mut ephemeral = EphemeralAccount::new_from_frozen(frozen_snapshot, signer_id, public_key)?;
+    let mut results = Vec::with_capacity(group.len());
+    for (signed_tx, cost) in group.iter() {
+        let vr = ephemeral.apply_transaction_ephemeral(
+            config,
+            signed_tx,
+            cost,
+            block_height,
+            current_protocol_version,
+        )?;
+        results.push((signed_tx.clone(), vr));
+    }
+    ephemeral.commit(state_update, signer_id, public_key);
```

Comment:
I suggest that the results should be committed to `TrieUpdate` in the main thread *after* rayon is done with verifications. Otherwise you're gonna hit the original problem of needing to mutate `TrieUpdate` from multiple threads which isn't gonna be easy.